File: firmware/domain/src/backup.rs
//! Versioned backup envelope: validate → preview → commit (DATA-002/003).
//!
//! A backup is portable JSON containing restorable user configuration and
//! calibration but never network credentials. Restore validates the whole
//! document against a checksum and schema version before committing; the
//! previous committed state remains readable until the replacement
//! transaction succeeds.

use crate::store::StorageError;
use heapless::String;

pub const BACKUP_SCHEMA_VERSION: &str = "0.1";
/// Bounded backup document for host tests and staging.
#[derive(Clone, Debug, Default)]
pub struct BackupDoc {
    pub schema_version: String<8>,
    /// Raw JSON payload for calibration, exactly as stored.
    pub calibration_json: String<512>,
    pub spool_json: String<256>,
    pub events_json: String<4096>,
    /// Integrity checksum of the canonical payload (FNV-1a, accidental
    /// corruption only — not authenticity).
    pub checksum: u32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BackupError {
    UnsupportedVersion,
    ChecksumMismatch,
    Malformed,
    Storage(StorageError),
    TooLarge,
}

/// Canonical payload bytes used for checksumming; field order is fixed so
/// the checksum is reproducible across host and target.
pub fn canonical_payload(doc: &BackupDoc, out: &mut String<8192>) -> Result<(), BackupError> {
    use core::fmt::Write;
    write!(
        out,
        "v{}|cal:{}|spool:{}|events:{}",
        doc.schema_version, doc.calibration_json, doc.spool_json, doc.events_json
    )
    .map_err(|_| BackupError::TooLarge)?;
    Ok(())
}

pub fn compute_checksum(doc: &BackupDoc) -> Result<u32, BackupError> {
    let mut payload: String<8192> = String::new();
    canonical_payload(doc, &mut payload)?;
    Ok(crate::store::fnv1a(payload.as_bytes()))
}

/// Validate an inbound backup document without mutating anything.
pub fn validate(doc: &BackupDoc) -> Result<(), BackupError> {
    if doc.schema_version != BACKUP_SCHEMA_VERSION {
        return Err(BackupError::UnsupportedVersion);
    }
    let actual = compute_checksum(doc)?;
    if actual != doc.checksum {
        return Err(BackupError::ChecksumMismatch);
    }
    // Reject payloads with unbalanced JSON braces at the envelope level.
    for field in [
        doc.calibration_json.as_str(),
        doc.spool_json.as_str(),
        doc.events_json.as_str(),
    ] {
        let opens = field.matches('{').count();
        let closes = field.matches('}').count();
        if opens != closes {
            return Err(BackupError::Malformed);
        }
    }
    Ok(())
}

/// A validated restore that can be committed later, bound to the exact
/// document digest observed at validation time.
#[derive(Clone, Debug)]
pub struct StagedRestore {
    pub doc: BackupDoc,
    pub digest: u32,
}

/// Validate and stage a restore. The caller must present the same digest
/// again to `commit` (fresh confirmation bound to the validated digest).
pub fn stage(doc: &BackupDoc) -> Result<StagedRestore, BackupError> {
    validate(doc)?;
    Ok(StagedRestore {
        doc: doc.clone(),
        digest: doc.checksum,
    })
}

/// Commit a staged restore only if the confirmation digest matches the
/// staged digest exactly (single-use).
pub fn commit(
    staged: &mut Option<StagedRestore>,
    confirm_digest: u32,
) -> Result<BackupDoc, BackupError> {
    let Some(current) = staged.as_ref() else {
        return Err(BackupError::Malformed);
    };
    if current.digest != confirm_digest {
        return Err(BackupError::ChecksumMismatch);
    }
    let taken = staged.take().ok_or(BackupError::Malformed)?;
    Ok(taken.doc)
}

#[cfg(test)]
mod tests {
    use super::{
        BACKUP_SCHEMA_VERSION, BackupDoc, BackupError, String, commit, compute_checksum, stage,
        validate,
    };

    fn doc() -> BackupDoc {
        let mut d = BackupDoc {
            schema_version: String::try_from(BACKUP_SCHEMA_VERSION).unwrap(),
            calibration_json: String::try_from("{\"state\":\"valid\"}").unwrap(),
            spool_json: String::try_from("{\"empty_mass_g\":430.0}").unwrap(),
            events_json: String::try_from("{\"schema_version\":\"0.1\",\"events\":[]}").unwrap(),
            checksum: 0,
        };
        d.checksum = compute_checksum(&d).unwrap();
        d
    }

    #[test]
    fn valid_document_passes() {
        assert_eq!(validate(&doc()), Ok(()));
    }

    #[test]
    fn wrong_version_rejected() {
        let mut d = doc();
        d.schema_version = String::try_from("9.9").unwrap();
        // Recompute checksum so version check is what fails, not checksum.
        d.checksum = compute_checksum(&d).unwrap();
        assert_eq!(validate(&d), Err(BackupError::UnsupportedVersion));
    }

    #[test]
    fn tampered_payload_rejected() {
        let mut d = doc();
        d.spool_json = String::try_from("{\"empty_mass_g\":1.0}").unwrap();
        assert_eq!(validate(&d), Err(BackupError::ChecksumMismatch));
    }

    #[test]
    fn unbalanced_json_rejected() {
        let mut d = doc();
        d.calibration_json = String::try_from("{\"state\":\"valid\"").unwrap();
        d.checksum = compute_checksum(&d).unwrap();
        assert_eq!(validate(&d), Err(BackupError::Malformed));
    }

    #[test]
    fn restore_requires_matching_digest_and_is_single_use() {
        let mut staged = Some(stage(&doc()).unwrap());
        // Wrong digest fails and keeps the stage.
        assert_eq!(
            commit(&mut staged, 0xDEAD).err(),
            Some(BackupError::ChecksumMismatch)
        );
        assert!(staged.is_some());
        // Correct digest commits and consumes the stage.
        let digest = staged.as_ref().unwrap().digest;
        let committed = commit(&mut staged, digest).unwrap();
        assert_eq!(committed.schema_version, BACKUP_SCHEMA_VERSION);
        // Second commit has nothing to commit.
        assert!(commit(&mut staged, digest).is_err());
    }
}

File: app/src/backup.ts
// Backup validate -> preview -> commit helpers (DATA-003).
//
// Parity with firmware/domain/src/backup.rs: canonical payload is the
// exact byte string `v{schema}|cal:{calibration_json}|spool:{spool_json}
// |events:{events_json}` (see canonical_payload there) hashed with FNV-1a
// over UTF-8 bytes. Validation performs no mutation; the caller commits
// only with a matching digest.

export const BACKUP_SCHEMA_VERSION = "0.1";

export interface BackupDoc {
  schema_version: string;
  calibration_json: string;
  spool_json: string;
  events_json: string;
  checksum: number;
}

export class BackupError extends Error {
  constructor(
    public code:
      | "unsupported_version"
      | "checksum_mismatch"
      | "malformed"
      | "credential_present",
    message: string,
  ) {
    super(message);
    this.name = "BackupError";
  }
}

/** FNV-1a 32-bit over UTF-8 bytes; must match firmware store::fnv1a. */
export function fnv1a(text: string): number {
  let hash = 0x811c9dc5;
  for (const byte of new TextEncoder().encode(text)) {
    hash ^= byte;
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash >>> 0;
}

export function canonicalPayload(doc: BackupDoc): string {
  return `v${doc.schema_version}|cal:${doc.calibration_json}|spool:${doc.spool_json}|events:${doc.events_json}`;
}

export function computeChecksum(doc: BackupDoc): number {
  return fnv1a(canonicalPayload(doc));
}

/** Wi-Fi credentials must never appear in backups (protocol rule 6). */
const CREDENTIAL_KEYS = ["psk", "password", "passphrase", "ssid", "credential"];

function containsCredentialKey(raw: string): boolean {
  const lowered = raw.toLowerCase();
  return CREDENTIAL_KEYS.some((key) =>
    new RegExp(`"${key}"\\s*:`).test(lowered),
  );
}

/** A preview prepared by validation; commit requires the same digest. */
export interface StagedRestore {
  doc: BackupDoc;
  digest: number;
  preview: {
    calibration_state: string | null;
    spool_empty_mass_g: number | null;
    event_count: number;
  };
}

/**
 * Validate a backup without mutating local state. Throws BackupError on
 * any rule violation; the previous committed state stays in place.
 */
export function validateBackup(doc: unknown): StagedRestore {
  if (typeof doc !== "object" || doc === null) {
    throw new BackupError("malformed", "Backup is not an object");
  }
  const candidate = doc as Partial<BackupDoc>;
  if (
    typeof candidate.schema_version !== "string" ||
    typeof candidate.calibration_json !== "string" ||
    typeof candidate.spool_json !== "string" ||
    typeof candidate.events_json !== "string" ||
    typeof candidate.checksum !== "number"
  ) {
    throw new BackupError("malformed", "Backup envelope has missing fields");
  }
  if (candidate.schema_version !== BACKUP_SCHEMA_VERSION) {
    throw new BackupError(
      "unsupported_version",
      `Backup schema ${candidate.schema_version} is not supported`,
    );
  }
  const doc_: BackupDoc = {
    schema_version: candidate.schema_version,
    calibration_json: candidate.calibration_json,
    spool_json: candidate.spool_json,
    events_json: candidate.events_json,
    checksum: candidate.checksum,
  };
  if (computeChecksum(doc_) !== doc_.checksum >>> 0) {
    throw new BackupError(
      "checksum_mismatch",
      "Backup integrity check failed; the file may be corrupted",
    );
  }
  for (const field of [
    doc_.calibration_json,
    doc_.spool_json,
    doc_.events_json,
  ]) {
    const opens = (field.match(/{/g) ?? []).length;
    const closes = (field.match(/}/g) ?? []).length;
    if (opens !== closes) {
      throw new BackupError("malformed", "Backup payload has unbalanced JSON");
    }
    if (containsCredentialKey(field)) {
      throw new BackupError(
        "credential_present",
        "Backup contains credential-like fields and was rejected",
      );
    }
  }
  let calibrationState: string | null = null;
  let spoolEmpty: number | null = null;
  let eventCount = 0;
  try {
    const cal = JSON.parse(doc_.calibration_json) as { state?: unknown };
    calibrationState = typeof cal.state === "string" ? cal.state : null;
    const spool = JSON.parse(doc_.spool_json) as { empty_mass_g?: unknown };
    spoolEmpty =
      typeof spool.empty_mass_g === "number" ? spool.empty_mass_g : null;
    const events = JSON.parse(doc_.events_json) as { events?: unknown };
    eventCount = Array.isArray(events.events) ? events.events.length : 0;
  } catch {
    throw new BackupError("malformed", "Backup payload is not valid JSON");
  }
  return {
    doc: doc_,
    digest: doc_.checksum >>> 0,
    preview: {
      calibration_state: calibrationState,
      spool_empty_mass_g: spoolEmpty,
      event_count: eventCount,
    },
  };
}

/**
 * Commit a staged restore. The confirm digest must match the digest the
 * user reviewed; a mismatch means stale confirmation and changes nothing.
 */
export function commitRestore(
  staged: StagedRestore | null,
  confirmDigest: number,
): BackupDoc | null {
  if (staged === null) return null;
  if (staged.digest !== confirmDigest) return null;
  return staged.doc;
}

export function serializeBackup(doc: BackupDoc): string {
  return JSON.stringify(doc, null, 2);
}

export function parseBackup(text: string): unknown {
  return JSON.parse(text) as unknown;
}
